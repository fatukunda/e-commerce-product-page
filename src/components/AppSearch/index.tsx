import React, { FunctionComponent, useEffect, useState } from "react";
import { useAppDispatch } from "src/store/hooks";
import {
  setSearchActive,
  setProducts,
  setIsSearching,
} from "src/store/slices/searchSlice";
import { searchProduct } from "src/services/searchService";

const AppSearch: FunctionComponent = (): JSX.Element => {
  const [searchTerm, setSearchTerm] = useState("");
  const dispatch = useAppDispatch();
  const changeHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
    let term = event.target.value;
    setSearchTerm(term);
  };
  useEffect(() => {
    if (searchTerm.length > 1) {
      dispatch(setSearchActive(true));
      const fetchItems = async () => {
        dispatch(setIsSearching(true));
        const res = await searchProduct(searchTerm);
        dispatch(setIsSearching(false));
        dispatch(setProducts(res));
      };
      fetchItems();
    } else {
      dispatch(setSearchActive(false));
    }
  }, [searchTerm, dispatch]);
  return (
    <form>
      <div className="relative">
        <div className="flex absolute inset-y-0 left-0 items-center pl-3 pointer-events-none">
          <svg
            className="w-5 h-5 text-gray-500 dark:text-gray-400"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
            ></path>
          </svg>
        </div>
        <input
          title="search-input"
          type="search"
          id="default-search"
          className="block p-4 pl-10 w-full text-md text-tertiary-light bg-gray-50 rounded-lg border border-gray-300 focus:ring-0 focus:outline-none"
          placeholder="Search Products"
          value={searchTerm}
          onChange={changeHandler}
          required
        />
      </div>
    </form>
  );
};

export default AppSearch;
