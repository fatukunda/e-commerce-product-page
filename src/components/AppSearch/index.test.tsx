import { render, cleanup, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider as ReduxProvider } from "react-redux";
import AppSearch from "./";
import mockStore from "../mocks/store";

const renderAppSearch = () =>
  render(
    <ReduxProvider store={mockStore()}>
      <AppSearch />
    </ReduxProvider>
  );
describe("Tests for <AppSearch/>", () => {
  afterEach(cleanup);

  it("renders without crashing", () => {
    renderAppSearch();
  });
  test("AppSearch: Snapshot", () => {
    const component = renderer.create(
      <ReduxProvider store={mockStore()}>
        <AppSearch />
      </ReduxProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Searches for product"', () => {
    const { getByTitle } = renderAppSearch();
    const searchInput = getByTitle("search-input") as HTMLInputElement;
    fireEvent.change(searchInput, { target: { value: "prev" } });
    expect(searchInput.value).toBe("prev");
  });
});
