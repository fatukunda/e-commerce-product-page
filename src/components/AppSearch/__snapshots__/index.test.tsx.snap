// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <AppSearch/> AppSearch: Snapshot 1`] = `
<form>
  <div
    className="relative"
  >
    <div
      className="flex absolute inset-y-0 left-0 items-center pl-3 pointer-events-none"
    >
      <svg
        className="w-5 h-5 text-gray-500 dark:text-gray-400"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
        />
      </svg>
    </div>
    <input
      className="block p-4 pl-10 w-full text-md text-tertiary-light bg-gray-50 rounded-lg border border-gray-300 focus:ring-0 focus:outline-none"
      id="default-search"
      onChange={[Function]}
      placeholder="Search Products"
      required={true}
      title="search-input"
      type="search"
      value=""
    />
  </div>
</form>
`;
